pac4j-kerberos/src/main/java/org/pac4j/kerberos/credentials/extractor/KerberosExtractor.java

```diff
@@ -0,0 +1,44 @@
+package org.pac4j.kerberos.credentials.extractor;
+
+import org.pac4j.core.context.HttpConstants;
+import org.pac4j.core.context.WebContext;
+import org.pac4j.core.credentials.extractor.CredentialsExtractor;
+import org.pac4j.core.exception.CredentialsException;
+import org.pac4j.core.exception.HttpAction;
+import org.pac4j.kerberos.credentials.KerberosCredentials;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+
+/**
+ * To extract Kerberos headers.
+ *
+ * @author Garry Boyce
+ * @since 2.1.0
+ */
+public class KerberosExtractor implements CredentialsExtractor<KerberosCredentials> {
+    private final String clientName;
+
+    public KerberosExtractor(final String clientName) {
+        this.clientName = clientName;
+    }
+
+    @Override
+    public KerberosCredentials extract(WebContext context) throws CredentialsException, HttpAction {
+        final String header = context.getRequestHeader(HttpConstants.AUTHORIZATION_HEADER);
+        if (header == null) {
+            throw HttpAction.unauthorizedNegotiate("Kerberos Header not found", context);
```

Comment:
Yes, you're completely right. The direct client is meant to take the credentials provided with the request and not break the flow if there are no credentials. So you should have no `throw HttpAction.unauthorizedNegotiate` in the direct client.

To keep the authenticated user profile in session, the indirect client seems to be the right choice, but the indirect client also redirects to the identity server for login (at the beginning of the flow). In the case of basic auth, the login URL is the callback URL (where credentials are expected) and in that case, if credentials are missing, a 401 error is returned.

My guess is that you should have something really really close to the basic auth support: `DirectBasicAuthClient`, `BasicAuthExtractor` and `IndirectBasicAuthClient`.

The `KerberosExtractor` should be shared among both clients, should not throw any `HttpAction` while the `KerberosIndirectClient` should handle that.

My only question is whether a direct client makes sense for Kerberos, but you know better...